File: frontend/src/app/shared/models/attributes/caster-type.ts
import {SpellSlots} from '../spell-slots';
import {Attribute} from './attribute';
import {ListObject} from '../list-object';
import {AttributeType} from './attribute-type.enum';

export class CasterType extends Attribute {
  type = 'CasterType';
  multiClassWeight: number;
  roundUp: boolean;
  spellSlots: SpellSlots[];

  constructor(levels: ListObject[]) {
    super();
    this.attributeType = AttributeType.CASTER_TYPE;
    this.multiClassWeight = 0;
    this.roundUp = false;
    this.spellSlots = [];

    if (levels != null) {
      for (let i = 0; i < levels.length; i++) {
        const level = new ListObject(levels[i].id, levels[i].name, levels[i].sid);
        this.spellSlots.push(new SpellSlots(level));
      }
    }
  }
}

File: backend/src/main/java/com/herd/squire/models/attributes/CasterType.java
package com.herd.squire.models.attributes;

import com.herd.squire.models.SpellSlots;

import java.util.ArrayList;
import java.util.List;

public class CasterType extends Attribute {
    private List<SpellSlots> spellSlots;
    private int multiClassWeight;
    private boolean roundUp;

    public CasterType() {}

    public CasterType(String id, String name, String description, int sid, boolean author, int version, int multiClassWeight, boolean roundUp) {
        super(id, name, description, AttributeType.CASTER_TYPE, sid, author, version);
        this.spellSlots = new ArrayList<>();
        this.multiClassWeight = multiClassWeight;
        this.roundUp = roundUp;
    }

    public List<SpellSlots> getSpellSlots() {
        return spellSlots;
    }

    public void setSpellSlots(List<SpellSlots> spellSlots) {
        this.spellSlots = spellSlots;
    }

    public int getMultiClassWeight() {
        return multiClassWeight;
    }

    public void setMultiClassWeight(int multiClassWeight) {
        this.multiClassWeight = multiClassWeight;
    }

    public boolean isRoundUp() {
        return roundUp;
    }

    public void setRoundUp(boolean roundUp) {
        this.roundUp = roundUp;
    }
}
